src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -7858,56 +7875,258 @@ def apply(
-            self._modin_frame.data_column_snowflake_quoted_identifiers
-        )
+        if axis == 0:
+            frame = self._modin_frame
 
-        from snowflake.snowpark.modin.pandas.utils import try_convert_index_to_native
+            # To apply function to Dataframe with axis=0, we repurpose the groupby apply function by taking each
+            # column, as a series, and treat as a single group to apply function.  Then collect the column results to
+            # join together for the final result.
+            col_results = []
 
-        # current columns
-        column_index = try_convert_index_to_native(self._modin_frame.data_columns_index)
+            # If raw, then pass numpy ndarray rather than pandas Series as input to the apply function.
+            if raw:
 
-        # Extract return type from annotations (or lookup for known pandas functions) for func object,
-        # if not return type could be extracted the variable will hold None.
-        return_type = deduce_return_type_from_function(func)
+                def wrapped_func(*args, **kwargs):  # type: ignore
+                    raw_input_obj = args[0].to_numpy()
+                    args = (raw_input_obj,) + args[1:]
+                    return func(*args, **kwargs)
 
-        # Check whether return_type has been extracted. If return type is not
-        # a Series, tuple or list object, compute df.apply using a vUDF. In this case no column expansion needs to
-        # be performed which means that the result of df.apply(axis=1) is always a Series object.
-        if return_type and not (
-            isinstance(return_type, PandasSeriesType)
-            or isinstance(return_type, ArrayType)
-        ):
-            return self._apply_udf_row_wise_and_reduce_to_series_along_axis_1(
-                func,
-                column_index,
-                input_types,
-                return_type,
-                udf_args=args,
-                udf_kwargs=kwargs,
-                session=self._modin_frame.ordered_dataframe.session,
-            )
+                agg_func = wrapped_func
+            else:
+                agg_func = func
+
+            # Accumulate indices of the column results.
+            col_result_indexes = []
+            # Accumulate "is scalar" flags for the column results.
+            col_result_scalars = []
+
+            for (
+                data_column_pandas_label,
+                data_column_snowflake_quoted_identifier,
+            ) in zip(
+                frame.data_column_pandas_labels,
+                frame.data_column_snowflake_quoted_identifiers,
+            ):
+                data_col_frame = InternalFrame.create(
+                    ordered_dataframe=frame.ordered_dataframe,
+                    data_column_snowflake_quoted_identifiers=[
+                        data_column_snowflake_quoted_identifier
+                    ],
+                    data_column_pandas_labels=[data_column_pandas_label],
```

Comment:
Can you add a comment about what this is doing? My understanding is you are maintaining the index but grabbing only the last data column in the frame? If that is the case, why do this instead of using `drop`?

Reply:
Added comment. This is iterating through each of the data columns of the original df, so data_column_pandas_label, data_column_snowflake_quoted_identifier are the current column, we then create a frame reduced down to this single data column to call groupby_apply with apply function on.  So it's not always the last data column.  The index is preserved because the apply function might access via the index.